### PR TITLE
Api demos

### PR DESCRIPTION
**Pull request by codejaeger**
## PR Checklist
If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [ ] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [ ] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [ ] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [ ] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [ ] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [ ] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #
Examples for Graph animation demos

**How did you address these issues with this PR? What methods did you use?**

### CODE REVIEW

File: examples/graph_animations/example1.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+**Graph initialization**
+```julia
+# Parameters - Graph object | is directed? | width | height | starting position
+ga = GraphAnimation(graph, true, 300, 300, O)
+```
+
+**Node registration**
+```julia
+nodes = [Object(@Frames(prev_start()+5, stop=100), GraphNode(i, drawNode; animate_on=:scale, fill_color="yellow", border_color="black", text=string(i), text_valign=:middle, text_halign=:center)) for i in range(1, 8; step=1)]
```
Comment (Wikunia) [inline examples/graph_animations/example1.md:32]: It's not clear how this is linked to `ga`

Comment (Wikunia) [reply examples/graph_animations/example1.md:32]: In general what does `ga` do if I need to create the graph nodes and edges myself?

File: examples/graph_animations/example1.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+The algorithm can be explained in two ways:
+
+**Using simple coloring**
+* Use a different fill color to convey the new nodes in the queue
+* Change the color of visited nodes permanently
+
+```julia
+enqueue!(Q, 1)
+# The frames argument is optional here.
+changeNodeProperty!(ga, 1, :fill_color, "green")
```
Comment (Wikunia) [inline examples/graph_animations/example1.md:93]: does this increase some kind of global frame counter?

File: examples/graph_animations/example2.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+*Note - Dynamic layout will be much more computationally intensive than static layout*
+
+**Adding or removing nodes and edges**
+
+New nodes and edges can be added to the canvas or existing nodes deleted after the graph creation.
+```julia
+removeNode!(ag, 2)
+addNode!(ag, 7)
+addEdge!(ag, 1, 7)
```
Comment (Wikunia) [inline examples/graph_animations/example2.md:53]: we might want to think about other names here to make clear that this has something to do with Javis. Maybe even just `Javis.add_edge!`

File: examples/graph_animations/example2.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+If frame management becomes an issue, it is possible to skip it completely and let Javis handle the frame management, again through the use of reasonable defaults. For example, skipping frames in `highlightNode` schedules the node highlighting after the end of the previous animation specified on the graph. This animation could be any of the graph utility functions with the exception of `changeNodeProperty` and `changeEdgeProperty` for which the starting frame is used as reference. To avail of this default option, the `default_keyframes=true` option needs to be passed in `create_graph`.
+
+Most utility functions like `changeNodeProperty` or `highlightNode` or `animate_*` use Javis actions underneath in the implementations which makes it easy to arrange them via frames.
+
+**Using `animate_neighbors`**
+
+Animating neighboring nodes and edges can be simplified by using the `animate_neighbors` utlity function
+
+```julia
+animate_neighbors(ag, 1; animate_node_on=(:fill_color, "green"), animate_edge_on=(:color, "red"))
```
Comment (Wikunia) [inline examples/graph_animations/example2.md:125]: can we have something like `(fill_color = "green", color = "red")` to provide several different things?

File: examples/graph_animations/example2.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+```
+This internally makes a call to `changeNodeProperty` and `changeEdgeProperty`. If the keyword arguments are not provided, simple switching highlighting is used for the animation. 
+
+**Using `animate_path`**
+
+After the traversal has been done and the shortest path found, show the path by animating it.
+
+```julia
+# change back the graph to its original form
+changeNodeProperty!(ag, (node)->true, :fill_color, "white")
```
Comment (Wikunia) [inline examples/graph_animations/example2.md:135]: the `(node)->true` syntax is quite weird. What does that do?

Comment (Wikunia) [reply examples/graph_animations/example2.md:135]: Here we might be able to also specify several changes like above?

File: examples/graph_animations/example3.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+     Dict(:weight=>1000, :neighbors=>[(4, 3)]),
+     Dict(:weight=>1000, :neighbors=>[])]
+```
+
+(*Visualisation coming up*)
+
+**Register graph in Javis**
+```julia
+wg = GraphAnimation(g, false, 300, 300, O; 
+                    node_attribute_fn=(g, node, attr) -> n(g, node, attr),
```
Comment (Wikunia) [inline examples/graph_animations/example3.md:37]: unclear to me what this does

**Review by Wikunia — COMMENTED** [review]
In general it looks like a good first step to me. Maybe you can clarify the small comments I made throughout for now and we can discuss this further on Tuesday.

File: examples/graph_animations/example1.md (hunk 1 above)
Comment (codejaeger) [reply examples/graph_animations/example1.md:32]: I thought of keeping a reference internally to the currently active graph, like the way there is a  `CURRENT_VIDEO`. This however is not completely free of problems and restricts the way nodes can be added to graphs. I can change it to `GraphNode(ga, i, drawNode); ...`.

When adding a node, the first important thing it needs to do is generate a new layout. This however, is not done if the `mode` is `static` in which case it is generated only once when the whole layout is known.
The second thing, an ordering is kept in the graph object as to which node/edge appeared first. This helps in the utility functions like `animate_neighbors` where instead of placing actions on all neighboring nodes in the same frame range this ordering can be used to organize them sequentially.

File: examples/graph_animations/example2.md (hunk 4 above)
Comment (codejaeger) [reply examples/graph_animations/example2.md:125]: Yes, I see no problem with that. But the drawing properties specified would be required to be used in the drawing function otherwise animation on that property won't show up.

File: examples/graph_animations/example2.md (hunk 5 above)
Comment (codejaeger) [reply examples/graph_animations/example2.md:135]: It is just a function with a fixed signature applying that property change to all nodes matching the criteria.

File: examples/graph_animations/example3.md (hunk 6 above)
Comment (codejaeger) [reply examples/graph_animations/example3.md:37]: Since the type of graph used is not known in this example, to fetch node properties like `weight` from the input graph this function will be used. These weights are then mapped to a certain drawing property like `radius`.

File: examples/graph_animations/example1.md (hunk 2 above)
Comment (codejaeger) [reply examples/graph_animations/example1.md:93]: Yes, that will be the case for now. This will create an action on the node based on the last animation created on the graph through the utility functions.
In the future, when it is possible to specify frames for an action based on an action from a different object. This counter would not be required anymore.

File: examples/graph_animations/example2.md (hunk 5 above)
Comment (Wikunia) [reply examples/graph_animations/example2.md:135]: Then maybe make it the default but that's good to know

File: examples/graph_animations/example3.md (hunk 6 above)
Comment (Wikunia) [reply examples/graph_animations/example3.md:37]: I was wondering why this doesn't return the normal inf instead of the string as well but that's good to know.

File: examples/graph_animations/example1.md (hunk 1 above)
Comment (Wikunia) [reply examples/graph_animations/example1.md:32]: I'm just wondering what `ga` does with the `graph` object itself when one has to add the edges and nodes later anyway and what would happen if I add more nodes than in my `graph` ?

Comment (Wikunia) [reply examples/graph_animations/example1.md:32]: And yeah I prefer `GraphNode(ga, i, drawNode` and optionally remove the `i` when one adds them in order anyway.

File: examples/graph_animations/example2.md (hunk 5 above)
Comment (Wikunia) [reply examples/graph_animations/example2.md:135]: Regarding that you might want to create the function declarations and docstrings next :wink:
